### PR TITLE
Counter cache

### PR DESCRIPTION
**Pull request by hw676018683**
Introduce child_counter_cache
It counts the number of direct children

close https://github.com/stefankroes/ancestry/issues/303

### CODE REVIEW

File: test/concerns/counter_cache_test.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+  def test_counter_cache_when_updating_parent
+    AncestryTestDatabase.with_model :depth => 2, :width => 2, :counter_cache => true do |model, roots|
+      parent1 = roots.first.first
+      parent2 = roots.last.first
+      child = parent1.children.first
+
+      assert_difference 'parent1.reload.children_count', -1 do
+        assert_difference 'parent2.reload.children_count', 1 do
+          child.update parent: parent2
```
Comment (kbrock) [inline test/concerns/counter_cache_test.rb]: update_attributes ?

**Comment by hw676018683** [comment]
ok

**Comment by vzharkov** [comment]
@hw676018683 any new for this PR?

**Comment by hw676018683** [comment]
@vzharkov  No

**Comment by vzharkov** [comment]
@hw676018683 May be you should resolve conflicts before PR merge?

**Comment by hw676018683** [comment]
Will it be merged if no any conflict? @kbrock  
cc @vzharkov

**Comment by mitchnick** [comment]
Thanks for creating and maintaining such an excellent gem! Looking forward to this PR, any idea when it will be ready?

File: lib/ancestry/instance_methods.rb [hunk 2]
```diff
@@ -93,6 +93,32 @@ def child_ancestry
...
+      end
+
+      parent_id && self.class.increment_counter(_counter_cache_column, parent_id)
+    end
+
+    def _counter_cache_column
+      self.ancestry_base_class.counter_cache_column.to_s
+    end
+
+    def ancestry_present?
```
Comment (kbrock) [inline lib/ancestry/instance_methods.rb]: this is not necessary - can use `ancestry?` instead or `parent?`

**Comment by kbrock** [comment]
ok. I'll circle back on this one. looks like `children?` can be tuned to take advantage of this too

**Comment by hw676018683** [comment]
CounterCache goes wrong sometimes, it may be better to keep current for `children?`

File: test/concerns/counter_cache_test.rb (hunk 1 above)
Comment (kbrock) [reply test/concerns/counter_cache_test.rb]: I'm still not happy about using `update` here

Comment (hw676018683) [reply test/concerns/counter_cache_test.rb]: but `update_attributes` is the alias of `update`

**Comment by kbrock** [comment]
Thank you @hw676018683 Sorry I sat on this for so long.

**Pull request merged**